data_scrapers/san_francisco.py

```diff
@@ -0,0 +1,268 @@
+        return super().request(f'{self.resource_url}{resource_id}', **kwargs)
+
+    def get_notes(self) -> str:
+        """
+        Get 'description' field of metadata for all resources. Collect into one string,
+        separated by 2 newlines.
+        """
+        meta_from_source = ''
+        for v in self.resource_ids.values():
+            url = f"{self.metadata_url}{v}.json"
+            data = super().request(url)
+            meta_from_source += data["description"] + '\n\n'
+        return meta_from_source
+
+    def get_update_times(self) -> List:
+        """
+        Return a list of update times for all resources.
+        """
+        update_times = []
+        for v in self.resource_ids.values():
+            url = f"{self.metadata_url}{v}.json"
+            data = super().request(url)
+            update_times.append(data["dataUpdatedAt"])
+        return update_times
+
+    def get_demographics(self) -> Dict:
+        """
+        Fetch cases by age, gender, race_eth. Fetch cases by transmission category
+        Returns the dictionary value for {"cases_totals": {}, "death_totals":{}}.
+        Note that SF does not provide death totals, so these datapoints will be -1.
+        To crete a DataFrame from the dictionary, run 'pd.DataFrame(get_demographics())'
+        """
+        # copy dictionary structure of global 'out' dictionary to local variable
+        demo_totals: Dict[str,Dict] = { "case_totals": dict(), "death_totals": dict() }
+        demo_totals["case_totals"]["gender"] = self.get_gender_table()
+        demo_totals["case_totals"]["age_group"] = self.get_age_table()
+        demo_totals["case_totals"]["transmission_cat"] = self.get_transmission_table()
+        demo_totals["case_totals"]["race_eth"] = self.get_race_eth_table()
+        return demo_totals
+
+    def get_timeseries(self) -> Dict[str,List[Dict]]:
+        """
+        Returns the dictionary value for "series": {"cases":[], "deaths":[], "tests":[]}.
+        To create a DataFrame from this dictionary, run
+        'pd.DataFrame(get_timeseries())'
+        """
+        out_series: Dict[str, List[Dict]] = {"cases": [], "deaths": [
+        ], "tests": []}  # dictionary structure for time_series
+        out_series["cases"] = self.get_cases_series()
+        out_series["deaths"] = self.get_deaths_series()
+        out_series["tests"] = self.get_tests_series()
+        return out_series
+
+    # Confirmed Cases and Deaths by Date and Transmission
+    # Note that cumulative totals are not directly reported, we are summing over the daily reported numbers
+    def get_cases_series(self) -> List[Dict]:
+        """Get cases timeseries json, sum over transmision cat by date"""
+        resource_id = self.resource_ids['cases_deaths_transmission']
+        params = {'case_disposition': 'Confirmed',
+                '$select': 'date,sum(case_count) as cases', '$group': 'date', '$order': 'date'}
+        data = self.request(resource_id, params=params)
```

Comment:
Ohhh, this is not really what I had in mind when I recommended a `SanFranciscoApi` class **(although it’s also totally fine; no need to do a whole bunch more refactoring at this point if it’s getting tedious)**.

I was trying to separate the concerns around “get me some raw data from the API” and “parse/reformat/reorganize the API’s data.” So you might have:

```py
# These functions are purely about requesting data from a remote API and
# deal only with network and parse-related things (e.g. determining URLs,
# retrying, parsing JSON, caching responses, etc.).
class SanFranciscoApi(SocrataApi):
    ...

    def get_cases_deaths_transmission(self, **kwargs):
        """
        Get the total cases by date, transmission type and disposition
        ('Death' or 'Confirmed').
        Browse this at: https://data.sfgov.org/COVID-19/COVID-19-Cases-Summarized-by-Date-Transmission-and/tvq9-ec9w
        """
        return self.request('tvq9-ec9w', **kwargs)

    # Or, a slightly more complex version that might simplify your usage.
    # Not sure if there are other requests to tvq9-ec9w that would be more
    # complex because of this.
    def get_cases_deaths_transmission(self, disposition, **kwargs):
        """
        Get the total cases by date, and transmission type for a given
        disposition ('Death' or 'Confirmed').
        Browse this at: https://data.sfgov.org/COVID-19/COVID-19-Cases-Summarized-by-Date-Transmission-and/tvq9-ec9w
        """
        params = {'case_disposition': disposition,
                  '$select': 'date,sum(case_count) as cases',
                  '$group': 'date',
                  '$order': 'date'}
        return self.request('tvq9-ec9w', params=params, **kwargs)


# In the same set of basic functions you had before:
def get_cases_series(api):
    data = api.get_cases_deaths_transmission({
        'case_disposition': 'Confirmed',
        '$select': 'date,sum(case_count) as cases',
        '$group': 'date',
        '$order': 'date'
    })
    # Or the fancier version:
    data = api.get_cases_deaths_transmission('Confirmed')

    # convert date from ISO string to 'yyyy-mm-dd'. convert number strings
    # to int. calculate daily cumulative.
    cases_series = []
    cumul = 0
    for entry in data:
        entry["date"] = entry["date"][0:10]
        entry["cases"] = int(entry["cases"])
        cumul += entry["cases"]
        entry["cumul_cases"] = cumul
        cases_series.append(entry)
    return cases_series
```

That probably doesn’t look like it’s saving you a whole lot, but it’s still pretty useful. You get function names that more clearly tell someone else reading your code just what’s happening, a better place to document issues around SF’s API itself, and a clear place to do other things that are particular to the API, like special parameter parsing, caching, HTTP headers, etc.

If you have a situation like Alameda, where you have both an API and some dashboards you’re scraping, you might have two separate clients to encapsulate the different concerns around making API requests vs. managing a webdriver instance.

The way you have it now, I’m not sure there’s a huge benefit from being methods on a `SanFranciscoApi` class rather than just creating an instance of `SocrataApi` and passing it around:

```py
def get_county():
    ...
    session = SocrataApi('https://data.sfgov.org/')
    ...
    get_cases_series(session)
    ...
```

(Having the class means you don’t have to pass the instance around, which is useful, but not a huge benefit.)

Anyway, what you have here is OK, too! Separating it out a bit more can help others understand or modify your code a bit more easily (or help you do so a couple weeks from know when you’ve forgotten all the details of what you did).